### PR TITLE
web: disable background feature flags evaluation

### PR DESCRIPTION
**Pull request by valerybugakov**
## Context

Based on [the discussion](https://sourcegraph.slack.com/archives/C03CSAER9LK/p1670260760447589) with @erzhtor, this PR disables the feature flags evaluation that was happening every minute for every feature in the background for each user. This behavior led to a huge volume of evaluate-feature-flags API requests. See [the Honeycomb dashboard](https://ui.honeycomb.io/sourcegraph/environments/dotcom/board/stqdV3uxYvo/web-app-HTTP-requests):

Screenshot 2022-12-13 at 15 05 22

## Preparation PRs

- https://github.com/sourcegraph/sourcegraph/pull/45604
- https://github.com/sourcegraph/sourcegraph/pull/45648

## How it affects our users 

Users will need to do a complete page refresh to update all feature flags. Also, feature flags are updated on re-renders once the `cacheTimeToLive` elapses.

## Test plan

1. CI is green.
2. Evaluate feature flag API requests are not initiated by the web application if a user doesn't do anything on the page.

## App preview:

- [Web](https://sg-web-vb-disable-feature-flags-refetches.onrender.com/search)

Check out the [client app preview documentation](https://docs.sourcegraph.com/dev/how-to/client_pr_previews) to learn more.

### CODE REVIEW

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.01% (-0.40 kb) | 0.19% (+28.10 kb) 🔺 | 0.24% (+28.51 kb) 🔺 | 0.13% (+1) 🔺 |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/vb/disable-feature-flags-refetches/compare-report.html;tab=live_object) for a full comparison between the commits b12f5c5 and a641750 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Comment by valerybugakov** [comment]
@erzhtor, could you take a look when you have a spare minute?

File: client/web/src/featureFlags/lib/FeatureFlagClient.ts [hunk 1]
```diff
@@ -56,12 +56,8 @@ export class FeatureFlagClient {
      */
     public get(flagName: FeatureFlagName): Observable<EvaluateFeatureFlagResult['evaluateFeatureFlag']> {
         if (!this.flags.has(flagName)) {
-            const flag$ = iif(
-                () => typeof this.refetchInterval === 'number' && this.refetchInterval > 0,
-                timer(0, this.refetchInterval),
-                timer(0)
-            ).pipe(
-                switchMap(() => fetchEvaluateFeatureFlag(this.requestGraphQLFunction, flagName).pipe(retry(3))),
+            const flag$ = fetchEvaluateFeatureFlag(this.requestGraphQLFunction, flagName).pipe(
```
Comment (erzhtor) [inline client/web/src/featureFlags/lib/FeatureFlagClient.ts:59]: While reviewing PR, I realized we could skip backend requests if we have a local override. So I think something like:
```ts
const flag$ = observeFeatureFlagLocalOverride(flagName).pipe(
    switchMap((overriddenValue) =>
        overriddenValue
            ? of(overriddenValue)
            : fetchEvaluateFeatureFlag(this.requestGraphQLFunction, flagName)
    ),
    distinctUntilChanged(),
    shareReplay(1)
);

```

File: client/web/src/featureFlags/useFeatureFlag.ts [hunk 2]
```diff
@@ -21,37 +23,44 @@ export function useFeatureFlag(flagName: FeatureFlagName, defaultValue = false):
...
-        )
+        })
 
         return () => {
             isMounted = false
-            setResult(({ value }) => ({ value, status: 'initial' }))
-            subscription.unsubscribe()
         }
-    }, [client, flagName])
+    })
```
Comment (erzhtor) [inline client/web/src/featureFlags/useFeatureFlag.ts:63]: Why did we change from observable subscription to updates on every render?

File: client/web/src/featureFlags/FeatureFlagsProvider.tsx [hunk 3]
```diff
@@ -51,65 +48,21 @@ const FeatureFlagsLocalOverrideAgent = React.memo(() => {
 })
 
 const MINUTE = 60000
```
Comment (erzhtor) [inline client/web/src/featureFlags/FeatureFlagsProvider.tsx:50]: I think we can increase this interval to 5-10 minutes.

**Review by erzhtor — COMMENTED** [review]
Thank you, @valerybugakov for tackling this! 💪  I left a few comments. Please, let me know if you have any questions.

File: client/web/src/featureFlags/useFeatureFlag.ts (hunk 2 above)
Comment (valerybugakov) [reply client/web/src/featureFlags/useFeatureFlag.ts:63]: Previously we subscribed to the feature flag and refetched in the background every minute. With this PR, we're removing background refetches in favor of the short-term in-memory cache. If the feature flag was updated and rerenders the component that relies on the feature, we want to show the updated UI (if cache TTL has elapsed already). That's why observable was replaced with `client.get()` calls on every render. Do you see any issues with this approach?

File: client/web/src/featureFlags/lib/FeatureFlagClient.ts (hunk 1 above)
Comment (valerybugakov) [reply client/web/src/featureFlags/lib/FeatureFlagClient.ts:66]: Addressed here https://github.com/sourcegraph/sourcegraph/pull/45647/commits/9af7b808ce4e2c5213e0c9e7dc6fe7b79858d844

File: client/web/src/featureFlags/FeatureFlagsProvider.tsx (hunk 3 above)
Comment (valerybugakov) [reply client/web/src/featureFlags/FeatureFlagsProvider.tsx:50]: Updated here https://github.com/sourcegraph/sourcegraph/pull/45647/commits/81dd0f4a8c466e361920178f3769f6c7ab5c331e

File: client/web/src/featureFlags/useFeatureFlag.ts (hunk 2 above)
Comment (valerybugakov) [reply client/web/src/featureFlags/useFeatureFlag.ts]: Migrated the `FeatureFlagsClient` to Promise API: https://github.com/sourcegraph/sourcegraph/pull/45647/commits/1db0e0189ad53cd4c8c4be695f775c55309ed392
Removed `useEffect` from the `useFeatureFlag` hook: https://github.com/sourcegraph/sourcegraph/pull/45647/commits/2cb4f7e8217be769845da3cbfba5fc97f2bae0b0

**Review by erzhtor — APPROVED** [review]
Overall, looks good to me! Thank you for handling it! 

One more feedback/concern is that with a new non-observable/reactive approach, there could be a part of the page with the old feature flag value and another with a new feature flag value. However, I guess that is an edge case. 
<details>

image

</details>

**Comment by valerybugakov** [comment]
> there could be a part of the page with the old feature flag value and another with a new feature flag value

Yeah, it's ok since users can get all the latest values together on a complete page refresh. If it causes confusion, we can go back to the polling solution.

**Pull request merged**